File: RoomLiveDataSampleJava/app/src/main/java/com/websarva/wings/android/roomlivedatasamplejava/MainActivity.java
package com.websarva.wings.android.roomlivedatasamplejava;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.websarva.wings.android.roomlivedatasamplejava.data.local.Cocktailmemo;
import com.websarva.wings.android.roomlivedatasamplejava.ui.MainViewModel;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
	private MainViewModel _mainViewModel;
	private CocktailListAdapter _cocktailListAdapter;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		ViewModelProvider provider = new ViewModelProvider(MainActivity.this);
		_mainViewModel = provider.get(MainViewModel.class);
		LiveData<Cocktailmemo> cocktailmemoLiveData = _mainViewModel.getCocktailmemoLiveData();
		cocktailmemoLiveData.observe(MainActivity.this, new CocktailmemoObserver());
		LiveData<List<Cocktailmemo>> cocktailmemoListLiveData = _mainViewModel.getCocktailmemoListLiveData();
		cocktailmemoListLiveData.observe(MainActivity.this, new CocktailmemoListObserver());

		RecyclerView rvCocktail = findViewById(R.id.rvCocktail);
		LinearLayoutManager layout = new LinearLayoutManager(MainActivity.this);
		rvCocktail.setLayoutManager(layout);
		DividerItemDecoration decoration = new DividerItemDecoration(MainActivity.this, layout.getOrientation());
		rvCocktail.addItemDecoration(decoration);
		_cocktailListAdapter = new CocktailListAdapter();
		rvCocktail.setAdapter(_cocktailListAdapter);

		Button btnSave = findViewById(R.id.btnSave);
		btnSave.setOnClickListener(new OnSaveButtonClick());
	}

	private class OnSaveButtonClick implements View.OnClickListener {
		@Override
		public void onClick(View view) {
			EditText etPrice = findViewById(R.id.etPrice);
			String price = etPrice.getText().toString();
			_mainViewModel.saveCocktailmemo(price);

		}
	}

	private class CocktailmemoViewHolder extends RecyclerView.ViewHolder {
		private final TextView _cocktailNameRow;
		private final TextView _cocktailPriceRow;
		public CocktailmemoViewHolder(View itemView) {
			super(itemView);
			_cocktailNameRow = itemView.findViewById(android.R.id.text1);
			_cocktailPriceRow = itemView.findViewById(android.R.id.text2);
		}

		public TextView getCocktailNameRow() {
			return _cocktailNameRow;
		}

		public TextView getCocktailPriceRow() {
			return _cocktailPriceRow;
		}
	}

	private class CocktailListAdapter extends RecyclerView.Adapter<CocktailmemoViewHolder> {
		private List<Cocktailmemo> _cocktailmemoList = new ArrayList<>();

		@NonNull
		@Override
		public CocktailmemoViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
			LayoutInflater inflater = LayoutInflater.from(MainActivity.this);
			View row = inflater.inflate(android.R.layout.simple_list_item_2, parent, false);
			row.setOnClickListener(new ListItemClickListener());
			CocktailmemoViewHolder holder = new CocktailmemoViewHolder(row);
			return holder;
		}

		@Override
		public void onBindViewHolder(@NonNull CocktailmemoViewHolder holder, int position) {
			Cocktailmemo cocktailmemo = _cocktailmemoList.get(position);
			TextView cocktailNameRow = holder.getCocktailNameRow();
			cocktailNameRow.setText(cocktailmemo.name);
			cocktailNameRow.setTag(cocktailmemo.id);
			TextView cocktailPriceRow = holder.getCocktailPriceRow();
			cocktailPriceRow.setText("¥" + cocktailmemo.price);
		}

		@Override
		public int getItemCount() {
			return _cocktailmemoList.size();
		}

		public void changeCocktailmemoList(List<Cocktailmemo> cocktailmemoList) {
			_cocktailmemoList = cocktailmemoList;
			notifyDataSetChanged();
		}
	}

	private class ListItemClickListener implements View.OnClickListener {
		@Override
		public void onClick(View view) {
			TextView cocktailNameRow = view.findViewById(android.R.id.text1);
			long id = (long) cocktailNameRow.getTag();
			String cocktailName = cocktailNameRow.getText().toString();
			_mainViewModel.prepareCocktailmemo(id, cocktailName);
		}
	}

	private class CocktailmemoListObserver implements Observer<List<Cocktailmemo>> {
		@Override
		public void onChanged(@NonNull List<Cocktailmemo> cocktailmemos) {
			_cocktailListAdapter.changeCocktailmemoList(cocktailmemos);
		}
	}

	private class CocktailmemoObserver implements Observer<Cocktailmemo> {
		@Override
		public void onChanged(@NonNull Cocktailmemo cocktailmemo) {
			TextView tvCocktailName = findViewById(R.id.tvCocktailName);
			tvCocktailName.setText(cocktailmemo.name);
			EditText etPrice = findViewById(R.id.etPrice);
			etPrice.setText(String.valueOf(cocktailmemo.price));
			Button btnSave = findViewById(R.id.btnSave);
			if(cocktailmemo.id == -1) {
				etPrice.setEnabled(false);
				btnSave.setEnabled(false);
			}
			else {
				etPrice.setEnabled(true);
				btnSave.setEnabled(true);
			}
		}
	}
}

File: RoomFlowSample/app/src/main/java/com/websarva/wings/android/roomflowsample/MainActivity.kt
package com.websarva.wings.android.roomflowsample

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.activity.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.websarva.wings.android.roomflowsample.data.local.Cocktailmemo
import com.websarva.wings.android.roomflowsample.ui.MainViewModel
import kotlinx.coroutines.flow.FlowCollector
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
	private val _mainViewModel by viewModels<MainViewModel>()
	private lateinit var _cocktailListAdapter: CocktailListAdapter

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_main)
		lifecycleScope.launch {
			repeatOnLifecycle(Lifecycle.State.STARTED) {
				_mainViewModel.cocktailmemoFlow.collect(CocktailmemoCollector())
			}
		}
		lifecycleScope.launch {
			repeatOnLifecycle(Lifecycle.State.STARTED) {
				_mainViewModel.cocktailmemoListFlow.collect(CocktailmemoListCollector())
			}
		}

		val rvCocktail = findViewById<RecyclerView>(R.id.rvCocktail)
		val layout = LinearLayoutManager(this@MainActivity)
		rvCocktail.layoutManager = layout
		val decoration = DividerItemDecoration(this@MainActivity, layout.orientation)
		rvCocktail.addItemDecoration(decoration)
		_cocktailListAdapter = CocktailListAdapter()
		rvCocktail.adapter = _cocktailListAdapter

		val btnSave = findViewById<Button>(R.id.btnSave)
		btnSave.setOnClickListener(OnSaveButtonClick())
	}

	private inner class OnSaveButtonClick : View.OnClickListener {
		override fun onClick(view: View?) {
			val etPrice = findViewById<EditText>(R.id.etPrice)
			val price = etPrice.text.toString()
			_mainViewModel.saveCocktailmemo(price)
		}
	}

	private inner class CocktailmemoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
		val cocktailNameRow: TextView
		val cocktailPriceRow: TextView
		init {
			cocktailNameRow = itemView.findViewById(android.R.id.text1)
			cocktailPriceRow = itemView.findViewById(android.R.id.text2)
		}
	}

	private inner class CocktailListAdapter() : RecyclerView.Adapter<CocktailmemoViewHolder>() {
		private var _cocktailmemoList: MutableList<Cocktailmemo> = mutableListOf()

		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CocktailmemoViewHolder {
			val inflater = LayoutInflater.from(this@MainActivity)
			val row = inflater.inflate(android.R.layout.simple_list_item_2, parent, false)
			row.setOnClickListener(ListItemClickListener())
			return CocktailmemoViewHolder(row)
		}

		override fun onBindViewHolder(holder: CocktailmemoViewHolder, position: Int) {
			val cocktailmemo = _cocktailmemoList.get(position)
			val cocktailNameRow = holder.cocktailNameRow
			cocktailNameRow.text = cocktailmemo.name
			cocktailNameRow.tag = cocktailmemo.id
			val cocktailPriceRow = holder.cocktailPriceRow
			cocktailPriceRow.text = "¥${cocktailmemo.price}"
		}

		override fun getItemCount(): Int {
			return _cocktailmemoList.size
		}

		fun changeCocktailmemoList(cocktailmemoList: MutableList<Cocktailmemo>) {
			_cocktailmemoList = cocktailmemoList
			notifyDataSetChanged()
		}
	}

	private inner class ListItemClickListener : View.OnClickListener {
		override fun onClick(view: View?) {
			view?.let {
				val cocktailNameRow = it.findViewById<TextView>(android.R.id.text1)
				val id = cocktailNameRow.tag as Long
				val cocktailName = cocktailNameRow.text.toString()
				_mainViewModel.prepareCocktailmemo(id, cocktailName)
			}
		}
	}

	private inner class CocktailmemoListCollector : FlowCollector<MutableList<Cocktailmemo>> {
		override suspend fun emit(cocktailmemos: MutableList<Cocktailmemo>) {
			_cocktailListAdapter.changeCocktailmemoList(cocktailmemos)
		}
	}

	private inner class CocktailmemoCollector : FlowCollector<Cocktailmemo> {
		override suspend fun emit(cocktailmemo: Cocktailmemo) {
			val tvCocktailName = findViewById<TextView>(R.id.tvCocktailName)
			tvCocktailName.text = cocktailmemo.name
			val etPrice = findViewById<EditText>(R.id.etPrice)
			etPrice.setText(cocktailmemo.price.toString())
			val btnSave = findViewById<Button>(R.id.btnSave)
			if(cocktailmemo.id == -1L) {
				etPrice.isEnabled = false
				btnSave.isEnabled = false
			}
			else {
				etPrice.isEnabled = true
				btnSave.isEnabled = true
			}
		}
	}
}
